File: backend/src/model/services/profiler_service.py
from typing import List, Tuple
from uuid import UUID
import requests
from model.services.To.user_stats import User_stats_to
from model.entities.user import User
from model.entities.collection import Collection
from model.daos.mongo_collection_dao import Mongo_collection_dao
from model.daos.mongo_user_dao import Mongo_user_dao
from model.exceptions import *
from os import environ


class Profiler_service():
    def __init__(self):
        self.url = environ.get("PROFILERS_SERVICE_URL")
        if self.url == None:
            raise RuntimeError("Env var PROFILERS_SERVICE_URL not declared.")
        self.algoritmos = ["modaresi", "grivas"]
        self.collection_dao = Mongo_collection_dao()
        self.user_dao = Mongo_user_dao()

    def profile_collection(self, filename, users_posts, algoritmo: str) -> Collection:
        if algoritmo not in self.algoritmos:
            raise NotSupportedAlgorithmException(algoritmo)

        try:
            response = requests.post(
                self.url+algoritmo, json=users_posts, timeout=1000)

            json = response.json()
            if response.status_code != 200:
                raise ServerNotAvailableException(json)

            users = [User(**user, posts=users_posts[user['user']])
                     for user in json['users']]
            coll = Collection(
                nombre=filename,
                algoritmo=algoritmo,
                tiempo=json['time'],
                users_stats=self.__calculate_user_stats(
                    users)
            )
            coll = self.collection_dao.create(coll)
            self.user_dao.create_collection_users(
                users, coll_id=coll.id)
            return coll

        except requests.exceptions.ConnectionError:
            raise ServerNotAvailableException(
                "The profiling server is not currently avaliable")
        except requests.exceptions.ReadTimeout:
            raise ServerTimeoutException(
                "The profiler took too long profiling the collection, try uploading a smaller one")

    def get_profiled_collection(self, collection_id: UUID) -> Collection:
        return self.collection_dao.get_collection(collection_id)

    def get_collection_users(self, collection_id: UUID, limit, offset, filters) -> Tuple[List[User], bool]:
        self.collection_dao.get_collection(collection_id)
        users, has_more = self.user_dao.get_collection_users(
            collection_id, limit=limit, skip=offset, filters=filters)
        return users, has_more

    def get_collection_stats(self, collection_id: UUID, filters) -> User_stats_to:
        self.collection_dao.get_collection(collection_id)
        filtered_users = self.user_dao.get_filtered_users(
            coll_id=collection_id, filters=filters)
        return User_stats_to(**self.__calculate_user_stats(filtered_users))

    def get_collections_list(self) -> list[Collection]:
        collections = self.collection_dao.get_collections()
        return collections

    def __calculate_user_stats(self, users):
        stats = {
            "total_users": len(users),
            "age": {
                '18-24': 0,
                '25-34': 0,
                '35-49': 0,
                '50-XX': 0
            },
            "gender": {
                "MALE": 0,
                "FEMALE": 0
            }
        }
        for user in users:
            stats['age'][user.age] += 1
            stats['gender'][user.gender] += 1

        return stats

File: frontend/src/services/ProfilerService.js
function handleErrors(response) {
    if (!response.ok) {
        throw Error("ProfileServiceError: " + response.statusText);
    }
    return response.json();
}

export default class ProfilerService {
    constructor() {
        throw new Error("Cannot instantiate abstract class");
    }

    static endpoint = "/api/profiler";

    static async profileUsers(file, algorithm) {
        const formData = new FormData();
        formData.append('file', file);
        formData.append('algoritmo', algorithm);
        return fetch(`${this.endpoint}/profile`, {
            method: 'POST',
            body: formData
        })
            .then(handleErrors)
            .then(this.#processColl)
            .catch(error => { throw error });
    }
    static async findCollections() {
        return fetch(`${this.endpoint}/collections`)
            .then(handleErrors)
            .then(collList => {
                return collList?.map(coll => {return this.#processColl(coll)})
            })
            .catch(error => { throw error });
    }
    static async getCollectionById(coll_id) {
        return fetch(`${this.endpoint}/collections/${coll_id}`)
            .then(handleErrors)
            .then(this.#processColl)
            .catch(error => { throw error });
    }

    static async findUsers(coll_id, limit = 100, offset = 0, filters = {}) {

        const queryParams = new URLSearchParams({
            limit: limit,
            offset: offset,
            ...this.#formatFilters(filters)
        });

        return fetch(`${this.endpoint}/collections/${coll_id}/users?${queryParams}`)
            .then(handleErrors)
            .then(data => {
                if (data.length == 0) {
                    return null
                }
                return data?.map(d => ({
                    id: d.id,
                    genero: d.gender[0] == 'M' ? 'Masculino' : 'Femenino',
                    edad: d.age.toLocaleLowerCase() == '50-xx' ? '50+' : d.age,
                    posts: d.posts,
                }))
            })
            .catch(error => { throw error });
    }
    static async getUsersStats(coll_id, filters = {}) {
        const queryParams = new URLSearchParams({
            ...this.#formatFilters(filters)
        });

        return fetch(`${this.endpoint}/collections/${coll_id}/stats?${queryParams}`)
            .then(handleErrors)
            .then(this.#getUsersStats)
            .catch(error => { throw error });
    }
    static #processColl = data => {
        return {
            id: data.id,
            name: data.nombre,
            algorithm: data.algoritmo,
            timestamp: data.fecha_creacion,
            time: data.tiempo,
            users: this.#getUsersStats(data.users_stats),
        }
    }
    static #formatFilters = filters => {
        const filtersTd = {}
        if (filters.age && filters.age == '50+') {
            filtersTd.age = '50-XX'
        }
        if (filters.gender) {
            filtersTd.gender = filters.gender == 'Masculino' ? 'MALE' : 'FEMALE'
        }
        return filtersTd;
    }
    static #getUsersStats = users_stats => {
        {
            return {
                'totalUsers': users_stats.total_users,
                'age': {
                    '18-24': users_stats['age']['18-24'],
                    '25-34': users_stats['age']['25-34'],
                    '35-49': users_stats['age']['35-49'],
                    '+50': users_stats['age']['50-XX'],
                },
                'gender': {
                    'Masculino': users_stats['gender']['MALE'],
                    'Femenino': users_stats['gender']['FEMALE'],
                }
            }
        }

    }
}
